server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java

```diff
@@ -0,0 +1,376 @@
+        long remotesPerSearchMax,
+        double remotesPerSearchAvg,
+        long skippedRemotes,
+        Map<String, Long> featureCounts,
+        Map<String, Long> clientCounts,
+        Map<String, PerClusterCCSTelemetry> byRemoteCluster
+    ) {
+        this.totalCount = totalCount;
+        this.successCount = successCount;
+        this.failureReasons = failureReasons;
+        this.took = took;
+        this.tookMrtTrue = tookMrtTrue;
+        this.tookMrtFalse = tookMrtFalse;
+        this.remotesPerSearchMax = remotesPerSearchMax;
+        this.remotesPerSearchAvg = remotesPerSearchAvg;
+        this.skippedRemotes = skippedRemotes;
+        this.featureCounts = featureCounts;
+        this.clientCounts = clientCounts;
+        this.byRemoteCluster = byRemoteCluster;
+    }
+
+    /**
+     * Creates a new empty stats instance, that will get additional stats added through {@link #add(CCSTelemetrySnapshot)}
+     */
+    public CCSTelemetrySnapshot() {
+        // Note this produces modifyable maps, so other snapshots can be added to it
+        failureReasons = new HashMap<>();
+        featureCounts = new HashMap<>();
+        clientCounts = new HashMap<>();
+        byRemoteCluster = new HashMap<>();
+        took = new LongMetricValue();
+        tookMrtTrue = new LongMetricValue();
+        tookMrtFalse = new LongMetricValue();
+    }
+
+    public CCSTelemetrySnapshot(StreamInput in) throws IOException {
+        this.totalCount = in.readVLong();
+        this.successCount = in.readVLong();
+        this.failureReasons = in.readMap(StreamInput::readLong);
+        this.took = LongMetricValue.fromStream(in);
+        this.tookMrtTrue = LongMetricValue.fromStream(in);
+        this.tookMrtFalse = LongMetricValue.fromStream(in);
+        this.remotesPerSearchMax = in.readVLong();
+        this.remotesPerSearchAvg = in.readDouble();
+        this.skippedRemotes = in.readVLong();
+        this.featureCounts = in.readMap(StreamInput::readLong);
+        this.clientCounts = in.readMap(StreamInput::readLong);
+        this.byRemoteCluster = in.readMap(PerClusterCCSTelemetry::new);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVLong(totalCount);
+        out.writeVLong(successCount);
+        out.writeMap(failureReasons, StreamOutput::writeLong);
+        took.writeTo(out);
+        tookMrtTrue.writeTo(out);
+        tookMrtFalse.writeTo(out);
+        out.writeVLong(remotesPerSearchMax);
+        out.writeDouble(remotesPerSearchAvg);
+        out.writeVLong(skippedRemotes);
+        out.writeMap(featureCounts, StreamOutput::writeLong);
+        out.writeMap(clientCounts, StreamOutput::writeLong);
+        out.writeMap(byRemoteCluster, StreamOutput::writeWriteable);
+    }
+
+    /**
+     * Add the provided stats to the ones held by the current instance, effectively merging the two
+     */
+    public void add(CCSTelemetrySnapshot stats) {
```

Comment:
It _looks_ like those adds will always be called in the single thread since the code in `TransportClusterStatsAction` collects node responses in a list and then passes it to final response ctor - so that's where the adds will happen, and it's single thread.